File: 03-SOLID-Principles/src/go/dip.go
package main

import "fmt"

// DIP — Dependency Inversion Principle
// OrderService depends on Database interface, not a concrete type

type Database interface {
	Save(data string)
}

type MySQLDB struct{}
func (MySQLDB) Save(data string) { fmt.Println("[MySQL] Saved:", data) }

type PostgreSQLDB struct{}
func (PostgreSQLDB) Save(data string) { fmt.Println("[PostgreSQL] Saved:", data) }

type InMemoryDB struct{}
func (InMemoryDB) Save(data string) { fmt.Println("[InMemory] Saved:", data) }

type OrderService struct {
	DB Database // depends on abstraction
}

func (o *OrderService) PlaceOrder(order string) {
	fmt.Println("Placing order:", order)
	o.DB.Save(order)
}

File: 03-SOLID-Principles/src/rust/dip.rs
/// DIP — Dependency Inversion Principle
/// OrderService depends on a Database trait, not a concrete type

pub trait Database {
    fn save(&self, data: &str);
}

pub struct MySQL;
impl Database for MySQL {
    fn save(&self, data: &str) { println!("[MySQL] Saved: {}", data); }
}

pub struct PostgreSQL;
impl Database for PostgreSQL {
    fn save(&self, data: &str) { println!("[PostgreSQL] Saved: {}", data); }
}

pub struct InMemoryDB;
impl Database for InMemoryDB {
    fn save(&self, data: &str) { println!("[InMemory] Saved: {}", data); }
}

pub struct OrderService<'a> {
    db: &'a dyn Database,   // depends on abstraction (trait object reference)
}

impl<'a> OrderService<'a> {
    pub fn new(db: &'a dyn Database) -> Self {
        OrderService { db }
    }

    pub fn place_order(&self, order: &str) {
        println!("Placing order: {}", order);
        self.db.save(order);
    }
}
